### PR TITLE
Thick Lines and Varying Point Sizes for Programmable Renderer

### LINKED ISSUE #8011: Drawing ofMesh with Vertex Colors is not cleared for next mesh draw without colors
Using the programmable renderer.
```
//--------------------------------------------------------------
void ofApp::draw(){
	ofSetColor(255, 255, 255);
	
	ofMesh boxMesh;
	boxMesh.setMode(OF_PRIMITIVE_LINES);
	
	float bsize = 120.f;
	boxMesh.addVertex(glm::vec3(-bsize, -bsize, -bsize));
	boxMesh.addVertex(glm::vec3(bsize, -bsize, -bsize));
	boxMesh.addVertex(glm::vec3(bsize, bsize, bsize));
	boxMesh.addVertex(glm::vec3(-bsize, bsize, bsize));
	
	boxMesh.addColor(ofColor::red);
	boxMesh.addColor(ofColor::yellow);
	boxMesh.addColor(ofColor::green);
	boxMesh.addColor(ofColor::blue);
	
	boxMesh.addIndex(0);
	boxMesh.addIndex(1);
	//
	boxMesh.addIndex(1);
	boxMesh.addIndex(2);
	//
	boxMesh.addIndex(2);
	boxMesh.addIndex(3);
	//
	boxMesh.addIndex(3);
	boxMesh.addIndex(0);
	
	
	
	ofPolyline circle;
	circle.arc(glm::vec3(0.f, 0.0f, 0.0f), 120.f, 120.f, 0, 360, true );
	
	ofPushMatrix(); {
		ofTranslate( ofGetWidth()/2, ofGetHeight()/2 );
		
		ofPushMatrix(); {
			ofTranslate(180, 0, 0);
//            boxMesh.drawWireframe();
			boxMesh.draw();
		} ofPopMatrix();
		
		ofPushMatrix(); {
			ofTranslate(-180, 0, 0);
			ofSetColor( 100.0f, (sinf(ofGetElapsedTimef()) * 0.5f + 0.5f) * 255, 50.0 );
			ofSetColor( ofColor::orange );
			circle.draw();
		} ofPopMatrix();
	} ofPopMatrix();
}
```
The above code outputs the following image
Screenshot 2024-06-04 at 9 20 54 PM
 
Commenting out the `boxMesh.addColor();` lines yields the following output
Screenshot 2024-06-04 at 9 21 08 PM

### PR DESCRIPTION
**Pull request by NickHardeman**
This PR attempts to mimic some of the original line and point drawing functionality allowed by earlier versions of GL for the Programmable GL Renderer pipeline while introducing some simple additional functionality. 

### Lines
This PR aims to enable varying line width for the following:
- ofMesh with mode  GL_LINES, GL_LINE_STRIP or GL_LINE_LOOP
- ofPolyline
- ofPath ( which draws the poly lines and sets the width from the stroke width on the path )
- ofDrawLine(x,y)

Currently supported attributes:
- Color per line point
- Textures (y coord is 0 -> tex height ) 
- Global color

The approach for the thick lines involves adding additional points per vertex to a vbo mesh that feeds additional VAO attributes to a shader.
- Allows for referencing next and previous vertex references
- Does not require a geometry shader ( not available in all instances or platforms ) 
- Renders 4 vertices per source vertex for mitered lines (GL_LINE_STRIP, GL_LINE_LOOP) and 2 vertices for GL_LINES 
- Can set screen space width (pixel width) regardless of distance to camera or world width ( size attenuation )

This PR currently does not intend to do the following
- Add end caps, round or otherwise
- Address self intersection

Example Code:

```
path.setStrokeWidth(mLineWidth);
path.draw();
```

```
ofSetLineWidth(mLineWidth);
mesh.setMode(OF_PRIMITIVE_LINES);
mesh.draw();
```

### Points
Points can vary in size based of a newly introduced function `ofSetPointSize(pointSize);` 
Supports
- Per point color
- Textures ( single texture for all points )
- Smoothing via a shader 

Example code:
```
ofSetPointSize(pointSize);
ofEnablePointSprites();
texture.bind();
mesh.drawVertices();
texture.unbind();
```

**_Core functionality to the GL Programmable Renderer has been changed and thus requires extensive testing._** 

**TODO:**
- [x] Optimizations
- [ ] Thorough testing
- [x] Cleaning up and comments
- [x] Update examples/gl/pointsAsTexturesExample
- [x] Create lines example ( in examples/graphics since Cairo also supports line width, not sure if it should be in examples/gl).

Fixes #8011

Screenshot 2024-06-13 at 10 53 04 PM

Screenshot 2024-06-15 at 9 16 29 PM

### CODE REVIEW

**Comment by dimitre** [comment]
Wowww marvellous work @NickHardeman

**Comment by ofTheo** [comment]
Yay! Wow, nice work @NickHardeman!

**Comment by NickHardeman** [comment]
This PR will close #8021

**Comment by NickHardeman** [comment]
When drawing an ofPolyline and a texture is bound, the tex coords get added so that the first point has texcoord.x == 0.0 and the last point has texcoord.x == texture.width. 
The y coord gets set between 0 and texture height based on the side of the line the dynamic points get added ( for poly lines and meshes ).
_**Should there be a way to control the texture coordinates for an ofPolyline?**_

Screenshot 2024-06-20 at 9 30 26 PM

**Comment by NickHardeman** [comment]
Without changing the vertices of miters, they can look like it does below, here they are clamped, but it doesn't look right with smaller angles between lines. Notice the colored lines.
Screenshot 2024-06-19 at 10 57 33 PM

**Comment by NickHardeman** [comment]
Some relevant discussions:
https://github.com/openframeworks/openFrameworks/issues/730
https://github.com/openframeworks/openFrameworks/issues/7689
Fixes #3460

File: libs/openFrameworks/gl/ofGLProgrammableRenderer.cpp [hunk 1]
```diff
@@ -1188,7 +1353,7 @@ void ofGLProgrammableRenderer::popStyle() {
 void ofGLProgrammableRenderer::setStyle(const ofStyle & style) {
 
 	//color
-	setColor((int)style.color.r, (int)style.color.g, (int)style.color.b, (int)style.color.a);
+	setColor(style.color.r, style.color.g, style.color.b, style.color.a);
```
Comment (danoli3) [inline libs/openFrameworks/gl/ofGLProgrammableRenderer.cpp:1356]: nice

File: libs/openFrameworks/gl/ofGLProgrammableRenderer.h [hunk 2]
```diff
@@ -293,13 +353,23 @@ class ofGLProgrammableRenderer: public ofBaseGLRenderer{
...
-	ofShader defaultNoTexNoColor;
+	
+	mutable GLuint mDrawMode = GL_TRIANGLES;
+	std::unordered_map<GLuint, LinesBundle> mLinesBundleMap;
+	mutable bool mBRenderingLines = false;
+	mutable bool mBLineSizeAttenutation = false; // screen space
+	mutable bool mBEnableLinesShaders = true;
+
+	//	the index GL_TRIANGLES store everything that is not GL_POINTS or GL_LINES, GL_LINE_STRIP
+	std::unordered_map<GLuint, std::shared_ptr<ShaderCollection> > mDefaultShadersMap;
```
Comment (danoli3) [inline libs/openFrameworks/gl/ofGLProgrammableRenderer.h:364]: cool idea

**Review by danoli3 — CHANGES_REQUESTED** [review]
Amazing, yeah just rebase onto latest master and lets check ci

**Comment by NickHardeman** [comment]
Thank you @danoli3. I can update next week when I am back at a computer 😀

**Comment by danoli3** [comment]
Wicked, we should just take a bunch of snapshots of gl / visual examples first draw frame:
- pre 
- with changes

**Review by ofTheo — APPROVED** [review]

**Comment by ofTheo** [comment]
I'd be good to merge this now - so we can get it in front of people. 
Especially as it's fixing something that didn't work before.

**Comment by danoli3** [comment]
I'm just running a regression comparer on this and then good to merge! So far collected images Pre

Just building opencv quality comparer for some analysis, saw opencv missing contrib modules, adding them now to build scripts 

I'll add comparer project to dev projects

**Comment by danoli3** [comment]
Looks good though overall good to merge

**Review by danoli3 — APPROVED** [review]

**Comment by ofTheo** [comment]
Thanks @danoli3 ! 

The difference between the rects in the graphics example is because they are animated randomly. 

And the difference in the color rects example looks more to do with color differences between the programmable renderer and the legacy renderer ( though why that is I am not sure ). 

I am guessing neither are due to this PR though. 

The rect fitting example also uses random rects ( in red ) - so the delta is just that too

**Pull request merged**